Use the quiz repository classes in the front-end

m1r0 opened the pull request:
Resolves #7223

## Proposed Changes
* Refactor quiz methods that were making direct calls to the comments table. They are now using the new quiz repository classes.
* Added unit tests for the methods that were not covered.

## Testing Instructions

1. Install the Query Monitor plugin.
1. Create a lesson with a quiz.
1. Go to the quiz front-end and check the Query Monitor queries. There should be no calls to the comments tables related to the quiz.
2. Submit the quiz and check again.
3. Grade the quiz and check again.

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [x] Acceptance criteria is met
- [x] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [x] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [x] Hooks (p6rkRX-1uS-p2) and functions are documented
- [x] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [x] New UIs match the designs
- [x] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [x] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [x] "Needs Documentation" label is added if this change requires updates to documentation
- [x] Known issues are created as new GitHub issues

m1r0 committed:
Refactor `Sensei_Quiz::get_user_answers_feedback` to use the quiz repositories

m1r0 committed:
Refactor `Sensei_Quiz::is_quiz_awaiting_grade_for_user` to use the quiz repositories

m1r0 committed:
Refactor `Sensei_Quiz::maybe_get_button_html_for_quiz_footer` to use the quiz repositories

m1r0 committed:
Fix linting issues

merkushin commented:
Wow. Looking at Query Monitor is... revealing.
For some reason we have too many duplicated queries to progress tables.
We need to use transients there, I think.
Otherwise, we may not get any performance benefits.

merkushin commented on includes/class-sensei-quiz.php:
```diff
@@ -2420,19 +2436,30 @@ private static function maybe_get_button_html_for_quiz_footer( $lesson_id = null
+		}
 
-		$is_complete = $lesson_status && in_array( $lesson_status->comment_approved, [ 'complete', 'graded', 'passed', 'failed' ], true );
+		$is_quiz_complete = in_array(
```
Maybe extract it as a method of quiz progress? (If it makes sense, of course.) Like this one:
https://github.com/Automattic/sensei/blob/trunk/includes/internal/student-progress/quiz-progress/models/class-quiz-progress-abstract.php#L213C25-L213C25

m1r0 replied:
I was debating with myself if it makes sense, but now that you mention it... I've added the helper method in 1baf3dbba931e6c1825faff8fc7467f3b5778036.

merkushin reviewed (dismissed):
Thanks Miro! Looks good and works as described 👍 

A tiny notice for people who may read it: we still have queries to comments table, it happens when we create or update progress/quiz entries.

merkushin commented:
@m1r0 did you try to run tests with table-reading enabled? :)

m1r0 committed:
Add a helper method for checking if the quiz is completed

m1r0 commented:
Thanks for the review @merkushin!

> For some reason we have too many duplicated queries to progress tables.

Yep, but that was somewhat expected. I've added a card for it https://github.com/Automattic/sensei/issues/7297 but I don't think we will have time for it during this stage.

> Otherwise, we may not get any performance benefits.

I guess that there should still be performance benefits on bigger sites. By default, MySQL caches duplicate queries so they should be pretty fast even if they are many.

> did you try to run tests with table-reading enabled? :)

Yes, and it doesn't look good. 😄 We have 40+ failing tests. I've added a card for it https://github.com/Automattic/sensei/issues/7296. Not sure if we do it at this stage. If we have the time, it would be nice to fix that.
